File: Data Structures & Algorithms/jump-game/submission-1.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        if n == 1:
            return True
        reachable = [0] * n

        reachable[0] = nums[0]
        for i in range( n ):
            furthest = max( i + nums[i], reachable[i] )
            # print(i,furthest)
            if reachable[i] == 0:
                return False
            if furthest >= n:
                return True

            for j in range( 1, nums[i]+1 ):
                reachable[i+j] = furthest
            # print(reachable)
            
        return reachable[-1] > 0

File: Data Structures & Algorithms/maximum-subarray/submission-1.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxSum = nums[0];
        int runningSum = nums[0];
        for( int i = 1; i < nums.size(); i++ ){
            int num = nums[i];
            if( runningSum <= 0){
                // case 1: runningSum is negative and
                // we have a number bigger than running sum (can be negative too)
                // start a new running sum
                if ( num > runningSum ){
                    runningSum = num;
                    maxSum = std::max(maxSum, runningSum);
                }
                // case 2: runningSum is negative and
                // our number is smaller; reset runningSum
                else {
                    runningSum = num;
                }
            }
            else {
                // case 3: runningSum is positive. adding this number does not make it negative
                if ( num + runningSum >= 0 ){
                    runningSum += num;
                    maxSum = std::max(maxSum, runningSum);
                }
                // case 4: adding this number makes runningSum negative; reset
                else {
                    runningSum = num;
                }
            }
        }
        return maxSum;
    }
};
